### DISCUSSION TITLE
NAT hole punching

### DISCUSSION THREAD

**Discussion by maboesanman** [n0-computer/iroh#1210] [General]
Does Iroh require all peers to port forward right now or is there a scheme for p2p nat hole punching?
Does it depend on whether or not the peer is publishing anything?

**Comment by maboesanman** [comment]
Upon playing around with iroh a little more, it appears this works quite nicely!

**Reply by dignifiedquire** [reply]
Glad to heare it. Are you using the version from `main` or a released version?

**Reply by maboesanman** [reply]
From main, and invoking with `cargo run --release -- <args>`

**Comment by dignifiedquire** [comment]
> Does Iroh require all peers to port forward right now or is there a scheme for p2p nat hole punching?

We are currently working on the first release that will include hole punching & relaying to ensure connectivity across the board. You can track our progress here: https://github.com/n0-computer/iroh/milestone/9

> Does it depend on whether or not the peer is publishing anything?

The easier the peer getting dialed (usually the one providing the data) is to reach, the easier connection establishment will be.

**Reply by maboesanman** [reply]
Development of iroh is happening faster than I realized!

Is peer discovery also in the works? In the future will I be able to invoke `iroh get <cid>` and be delivered my content from anyone who has it?
